android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardWebViewClient.java

```diff
@@ -0,0 +1,265 @@
+/**
+ * Copyright (C) 2023 SIL International. All rights reserved.
+ */
+package com.keyman.engine;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.graphics.Bitmap;
+import android.graphics.RectF;
+import android.net.Uri;
+import android.util.Log;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import android.widget.RelativeLayout;
+
+import com.keyman.engine.KeyboardEventHandler.EventType;
+import com.keyman.engine.KMManager.KeyboardType;
+import com.keyman.engine.util.KMLog;
+import com.keyman.engine.data.Keyboard;
+
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public final class KMKeyboardWebViewClient extends WebViewClient {
+  public static final String TAG = "KMKeyboardWebViewClient";
+
+  public static Context context;
+  private static KeyboardType keyboardType;
+  private static boolean keyboardLoaded;
+
+  KMKeyboardWebViewClient(Context context, KeyboardType keyboardType) {
+    KMKeyboardWebViewClient.context = context;
+    KMKeyboardWebViewClient.keyboardType = keyboardType;
+    KMKeyboardWebViewClient.keyboardLoaded = false;
+
+    if (keyboardType != KeyboardType.KEYBOARD_TYPE_INAPP && keyboardType != KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+      KMLog.LogError(TAG, String.format("Cannot initialize: Invalid keyboard type: %s", keyboardType.toString()));
+    }
+  }
+
+  public void setContext(Context context) {
+    KMKeyboardWebViewClient.context = context;
+  }
+
+  public boolean getKeyboardLoaded() {
+    return keyboardLoaded;
+  }
+
+  public void setKeyboardLoaded(boolean keyboardLoaded) {
+    KMKeyboardWebViewClient.keyboardLoaded = keyboardLoaded;
+  }
+
+  @Override
+  public void onPageStarted(WebView view, String url, Bitmap favicon) {
+  }
+
+  @Override
+  public void onPageFinished(WebView view, String url) {
+    Log.d("KMEA", "onPageFinished: [inapp] " + url);
+    shouldOverrideUrlLoading(view, url);
+  }
+
+  private void pageLoaded(WebView view, String url) {
+    Log.d("KMEA", "pageLoaded: [inapp] " + url);
+    if (KMManager.getKMKeyboard(keyboardType) == null) {
+      KMLog.LogError(TAG, "pageLoaded and InAppKeyboard null");
+      return;
+    }
+    KMManager.getKMKeyboard(keyboardType).keyboardSet = false;
```

Comment:
I refactored a lot of these `KMManager.getKMKeyboard()` calls in the follow-on #7993